### DISCUSSION TITLE
Jet energy correction uncertainty propagation trough selection

### DISCUSSION THREAD

**Discussion by JulesVandenbroeck** [columnflow/columnflow#441] [Q&A] [answered]
Hi everyone,

I was trying to understand how you can propagate uncertainties that change the selection can be propagated trough columnflow (eg: JEC, JER, electron scale and smearing*). Looking at previous commits about JEC and comments in the [analysis template](https://github.com/columnflow/columnflow/blob/master/analysis_templates/cms_minimal/__cf_module_name__/config/analysis___cf_short_name_lc__.py#L202) there used to be a flag "selection_dependent" that indicated this behavior. However in the current version I found no way to add a shift that the task `cf.SelectEvents` could recognize and did not overwrite to `nominal`.  Is there a method to run `cf.SelectEvents` with a shift in the current version of columnflow, and is there an example (this might be interesting to then also add to the example since it has dummy JEC uncertainties which are then not propagated trough.

Kind Regards,
Jules Vandenbroeck

*The electron scale and smearing are rather new uncertainty recommendations and I wrote a [code based on JEC's here](https://github.com/GhentAnalysis/columnflow/blob/origin/electron_ss/columnflow/calibration/cmsGhent/electron.py) (electron only for now) which could be used by other users.

ps: I opened an [issue](https://github.com/columnflow/columnflow/issues/436) last week on using `ROOT` in `cf.CalibrateEvents` which is not possible using a cmssw environment (missing gfal2).

**Comment by JulesVandenbroeck** [comment, marked as answer]
I found an answer to my own question:

shifts can be recognized by a `selector` only if they are included in `self.shifts`. This can be done by adding `shifts = {name of shift,}` or in a more flexible way below where `default_shifts` is the default selector that includes shifts that have the tag `selection dependent`.
```
# make sure to add register_selector_shifts = False to the selector variables
# if "register_selector_shifts" is True, include shifts with tag "selection_dependent"
@default.init
def default_init(self: Selector) -> None:
    if self.register_selector_shifts:
        self.shifts = {shift.name for shift in self.config_inst.shifts if "selection_dependent" in shift.tags}


default_shifts = default.derive("default_shifts", cls_dict={"register_selector_shifts": True})
```

**Comment by pkausw** [comment]
Hi @JulesVandenbroeck ,

glad to see that you were able to resolve this yourself! Just one more comment for future documentation:

In principlte, the `selection_dependent` tag is not strictly necessary, and also the `register_selector_shifts` is up to you of course (e.g. if you want one version of a selection that does consider a shift, and another that doesn't for debugging purposes). You can find examples of current implementations e.g. for

- an entry in the config: https://github.com/uhh-cms/hh2bbtautau/blob/master/hbt/config/configs_run2ul.py#L420-L445
- an example of where this is used: https://github.com/uhh-cms/hh2bbtautau/blob/master/hbt/selection/jet.py#L231
- the calibrator that employs the jec/jer corrections as they are currently implemented in columnflow: https://github.com/uhh-cms/hh2bbtautau/blob/master/hbt/calibration/default.py

Maybe this can add even more clarity, though I'm sure that your solution works perfectly fine 👍

**Reply by JulesVandenbroeck** [reply]
While applying the shift on selection worked and this discussion is closed I had one more point which can not be resolved in the same way. If you need to produce any variable or scale factor (eg b-tagging) that uses shifted columns like `Jet.pt `you need to propagate this change all the way to `cf.produceColumns`. There are two options for this that I could think of:

- Add all the variations (eg. "Jet.pt_jec_Total_up") to `cfg.x.keep_columns["cf.ReduceEvents"]` so that in `cf.produceColumns` when applying aliases it can access these columns. The keep_columns is however fixed in the config so that for each selection dependent shift you would need to save all the aliases of all other shifts.
- Add shifts to `cf.ReduceEvents` so that `local_shift` is not nominal and the Jet.pt is replaced with its alias in the reduced file. Adding shifts to cf.ReduceEvents I wasn't able to do in the same way as with `cf.SelectEvents` as there is no "Reducer" in the analysis. Since I found no reason to not always apply the aliases of the shift in `cf.ReduceEvents` (only happens when the shift is selection dependent), I changed the line `aliases = self.local_shift_inst.x("column_aliases", {})` to `aliases = self.global_shift_inst.x("column_aliases", {})` in [columnflow/tasks/reduction.py](https://github.com/columnflow/columnflow/blob/master/columnflow/tasks/reduction.py#L101).

If there is a simpler way to add shifts to cf.ReduceEvents so that `local_shift` would match `shift` this would solve the issue without having to change columnflow code and would be very helpful.

Kind Regards,
Jules

**Comment by mafrahm** [comment]
Hi @JulesVandenbroeck, thank you for bringing this to our attention! I just checked this today and can verify that indeed we do not store the varied jets when running with JEC/JER shifts applied. I think, your hotfix of using the `global_shift_inst` works fine, but we'll discuss this intern before applying the fix.

**Reply by mafrahm** [reply]
We just pushed a fix: https://github.com/columnflow/columnflow/commit/cb16bea064ec1d3689d6fe0cc44f1d6c619ec92c

Basically, ReduceEvents now registers the same shifts as SelectEvents, therefore the JER/JER uncertainties should now also be registered as `local_shift_inst`. I hope this solves your issues.
